### DISCUSSION TITLE
p2p behind NATs

### DISCUSSION THREAD

**Discussion by inosms** [libp2p/rust-libp2p#1970] [General]
Hi there!
Thanks for all the efforts you put into this library! I have a question and didn't exactly know where to ask it, thus this issue. 

I have been playing around with writing a decentralized backup tool and really want to try using this library for that. However I am stuck with the problem that **I don't know how to implement a peer to peer connection when both peers are behind different NATs** (using the IPFS Kademlia DHT for bootstrapping like in the ipfs-kad example).

I assume it is my lack of knowledge in p2p systems, but is there any chance you can hint me into the correct direction? I tried to reconstruct a go-libp2p example (https://github.com/libp2p/go-libp2p-examples/tree/master/chat-with-rendezvous), which lead me to #297 .  I was unable to translate calls to `discovery` like `discovery.Advertise()`. Would this be `add_provider()` in `rust_libp2p` (which is not yet implemented?)
Did I maybe misunderstand the purpose of this library?
Any help would be appreciated! Thanks! :)

(Also I think for other people who have the same/a similar use case, it would be good to have this kind of problem as an example in this repository - I will add that once I understand how to do it)

**Comment by tomaka** [comment]
:wave: 

If I'm not mistaken, `discovery.Advertise()` only advertises the current node to the Kademlia DHT, which is not exactly related to NAT traversal.

There are three mechanisms in play that are related to NATs:

- The `identify` behaviour, if used, will automatically query remote nodes for the external IP of the local node. This can then be retrieved through the `Swarm` and the `PollParameters`.
- There is a `relay` protocol and behaviour that can be used to relay communications between two nodes if they both connect a third. However this is not implemented (#725).
- It would eventually be nice to support UPnP or PCP, but this isn't implemented either (#558).

**Comment by inosms** [comment]
Thank you very much the detailed reply! This definitely makes things clearer! I will play around with it and see how far I can get 😃 
I assume this issue can be closed then.

**Comment by povilasb** [comment]
I remember asking @tomaka the same thing on IRC and he said you can use `rust-libp2p` with an existing system socket. So you can rely on external code to do NAT traversal and just use the TCP/UDP socket bound to address that is mapped to the externally reachable port...

At MaidSafe we have developed a pure Rust implementation solely for hole punching: https://github.com/ustulation/p2p . Unfortunately, it only works with mio, not tokio and you should glue it with DHT you decide to use, etc.

Wish there was an I/O free implementation of such library. Then it could be plugged into any type of networking implementation, etc. See the [hole punching](https://github.com/tox-rs/tox/blob/2a6d0884a2dadcdf09ff46c64c06a9df92fa2196/src/toxcore/dht/server/hole_punching.rs) module from tox-rs project :)

**Comment by tomaka** [comment]
I just wrote a summary of what we're doing at the moment: https://github.com/libp2p/rust-libp2p/issues/794#issuecomment-474782253
